lighthouse-core/gather/gatherers/source-maps.js

```diff
@@ -0,0 +1,126 @@
+
+/**
+ * This function is careful not to parse the response as JSON, as it will
+ * just need to be serialized again over the protocol, and source maps can
+ * be huge.
+ *
+ * @param {string} url
+ * @return {Promise<string|{errorMessage: string}>}
+ */
+/* istanbul ignore next */
+async function fetchSourceMap(url) {
+  try {
+    // eslint-disable-next-line no-undef
+    const response = await fetch(url);
+    return response.text();
+  } catch (err) {
+    return {errorMessage: err.toString()};
+  }
+}
+
+/**
+ * @fileoverview Gets JavaScript source maps.
+ */
+class SourceMaps extends Gatherer {
+  constructor() {
+    super();
+    /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+    this._scriptParsedEvents = [];
+    this.onScriptParsed = this.onScriptParsed.bind(this);
+  }
+
+  /**
+   * @param {Driver} driver
+   * @param {string} sourceMapUrl
+   * @return {Promise<{map: LH.Artifacts.RawSourceMap} | {errorMessage: string}>}
+   */
+  async fetchSourceMapInPage(driver, sourceMapUrl) {
+    driver.setNextProtocolTimeout(250);
+    /** @type {string|{errorMessage: string}} */
+    const sourceMapJsonOrError =
+      await driver.evaluateAsync(`(${fetchSourceMap})(${JSON.stringify(sourceMapUrl)})`);
+
+    // If object, then an error occurred.
+    if (typeof sourceMapJsonOrError === 'object') {
+      return sourceMapJsonOrError;
+    }
+
+    // Map was returned as JSON. See `fetchSourceMap` for why.
+    return {map: JSON.parse(sourceMapJsonOrError)};
+  }
+
+  /**
+   * @param {string} sourceMapURL
+   * @return {{map: LH.Artifacts.RawSourceMap}}
+   */
+  parseSourceMapFromDataUrl(sourceMapURL) {
+    const buffer = Buffer.from(sourceMapURL.split(',')[1], 'base64');
+    return {
+      map: JSON.parse(buffer.toString()),
+    };
+  }
+
+  /**
+   * @param {LH.Crdp.Debugger.ScriptParsedEvent} event
+   */
+  onScriptParsed(event) {
+    if (event.sourceMapURL) {
+      this._scriptParsedEvents.push(event);
+    }
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   */
+  async beforePass(passContext) {
+    const driver = passContext.driver;
+    driver.on('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.enable');
+  }
+
+  /**
+   * @param {LH.Gatherer.PassContext} passContext
+   * @return {Promise<LH.Artifacts['SourceMaps']>}
+   */
+  async afterPass(passContext) {
+    const driver = passContext.driver;
+
+    driver.off('Debugger.scriptParsed', this.onScriptParsed);
+    await driver.sendCommand('Debugger.disable');
+
+    /** @type {LH.Artifacts.SourceMap[]} */
+    const sourceMaps = [];
+    for (const event of this._scriptParsedEvents) {
+      const sourceMapURL = event.sourceMapURL;
+      if (!sourceMapURL) continue;
+
+      let sourceMapOrError;
+      try {
+        sourceMapOrError = sourceMapURL.startsWith('data:') ?
+          this.parseSourceMapFromDataUrl(sourceMapURL) :
+          await this.fetchSourceMapInPage(driver, sourceMapURL);
```

Comment:
@patrickhulce 	https://github.com/GoogleChrome/lighthouse/pull/9101/files/a6666fed5caf939e6a7816f1d457501314df8271#discussion_r290045041

Reply:
Can I fire all of them at once (with the same timeout) and count on them all returning? example, 10 source maps, will the browser make all 10 requests at once or will it make a handful and only issue new ones when one finishes?

Reply:
since these won't necessarily be small (so the overhead won't outweigh the savings), you could parallelize on the lighthouse side, using `Promise.all()` on the evaluateAsync calls.

> Can I fire all of them at once (with the same timeout) and count on them all returning? example, 10 source maps, will the browser make all 10 requests at once or will it make a handful and only issue new ones when one finishes?

the browser will max out how many connections it wants to make and queue the rest automatically.

That does make timeouts per request harder to deal with, though.

Reply:
Done.